halo2_proofs/src/dev/cost_model.rs

```diff
@@ -0,0 +1,364 @@
+            .collect();
+
+        let column_queries = queries.len();
+        queries.sort_unstable();
+        queries.dedup();
+        let point_sets = queries.len();
+
+        ModelCircuit {
+            k: opts.k,
+            max_deg,
+            advice_columns: opts.advice.len(),
+            lookups: opts.lookup.len(),
+            permutations: opts.permutation,
+            column_queries,
+            point_sets,
+        }
+    }
+}
+
+impl ModelCircuit {
+    /// Size of the proof in bytes
+    pub fn proof_size<const COMM: usize, const SCALAR: usize>(
+        &self,
+        comm_scheme: CommitmentScheme,
+    ) -> usize {
+        let size = |points: usize, scalars: usize| points * COMM + scalars * SCALAR;
+
+        // PLONK:
+        // - COMM bytes (commitment) per advice column
+        // - 3 * COMM bytes (commitments) + 5 * SCALAR bytes (evals) per lookup argument
+        // - COMM bytes (commitment) + 2 * SCALAR bytes (evals) per permutation argument
+        // - COMM bytes (eval) per column per permutation argument
+        let plonk = size(1, 0) * self.advice_columns
+            + size(3, 5) * self.lookups
+            + self
+                .permutations
+                .iter()
+                .map(|p| size(1, 2 + p.columns))
+                .sum::<usize>();
+
+        // Vanishing argument:
+        // - (max_deg - 1) * COMM bytes (commitments) + (max_deg - 1) * SCALAR bytes (h_evals)
+        //   for quotient polynomial
+        // - SCALAR bytes (eval) per column query
+        let vanishing = size(self.max_deg - 1, self.max_deg - 1) + size(0, self.column_queries);
+
+        // Multiopening argument:
+        // - f_commitment (COMM bytes)
+        // - SCALAR bytes (evals) per set of points in multiopen argument
+        let multiopen = size(1, self.point_sets);
+
+        let polycomm = match comm_scheme {
+            CommitmentScheme::IPA => {
+                // Polycommit IPA:
+                // - s_poly commitment (COMM bytes)
+                // - inner product argument (k rounds * 2 * COMM bytes)
+                // - a (SCALAR bytes)
+                // - xi (SCALAR bytes)
+                size(1 + 2 * self.k, 2)
+            }
+            CommitmentScheme::KZG => {
+                // Polycommit KZG:
+                // - quotient polynomial commitment (COMM bytes)
+                size(1, 0)
+            }
+        };
+
+        plonk + vanishing + multiopen + polycomm
+    }
+
+    /// Generate a report.
+    pub fn report<const COMM: usize, const SCALAR: usize>(
+        &self,
+        comm_scheme: CommitmentScheme,
+    ) -> String {
+        let mut str = String::new();
+        str.push_str(&format!("{:#?}", self));
+        str.push_str(&format!(
+            "Proof size: {} bytes",
+            self.proof_size::<COMM, SCALAR>(comm_scheme)
+        ));
+        str
+    }
+
+    /// Write a CSV report
+    pub fn report_csv<const COMM: usize, const SCALAR: usize, W: std::io::Write>(
```

Comment:
Now that @han0110 highlited this, can't you store the data in a HashMap style or similar and serialize it with `serde`??

That will make possible to have JSON, CSV or any other thing based on the client choice, not ours.

Reply:
Good idea. I'll use serde for reporting

Reply:
[5fb8194](https://github.com/privacy-scaling-explorations/halo2/pull/235/commits/5fb81946af23710b56e5749c5a2ae593a3f37883)